src/node_api.cc

```diff
@@ -0,0 +1,2430 @@
+      v8::Array::New(v8impl::V8IsolateFromJsEnv(e)));
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_array_with_length(napi_env e,
+                                          int length,
+                                          napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsValueFromV8LocalValue(
+      v8::Array::New(v8impl::V8IsolateFromJsEnv(e), length));
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_string_utf8(napi_env e,
+                                    const char* s,
+                                    int length,
+                                    napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  auto isolate = v8impl::V8IsolateFromJsEnv(e);
+  v8::Local<v8::String> str;
+  CHECK_NEW_FROM_UTF8_LEN(isolate, str, s, length);
+
+  *result = v8impl::JsValueFromV8LocalValue(str);
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_string_utf16(napi_env e,
+                                     const char16_t* s,
+                                     int length,
+                                     napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  auto isolate = v8impl::V8IsolateFromJsEnv(e);
+  auto str_maybe =
+      v8::String::NewFromTwoByte(isolate,
+                                 reinterpret_cast<const uint16_t*>(s),
+                                 v8::NewStringType::kInternalized,
+                                 length);
+  CHECK_MAYBE_EMPTY(str_maybe, napi_generic_failure);
+  v8::Local<v8::String> str = str_maybe.ToLocalChecked();
+
+  *result = v8impl::JsValueFromV8LocalValue(str);
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_number(napi_env e, double v, napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsValueFromV8LocalValue(
+      v8::Number::New(v8impl::V8IsolateFromJsEnv(e), v));
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_boolean(napi_env e, bool b, napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsValueFromV8LocalValue(
+      v8::Boolean::New(v8impl::V8IsolateFromJsEnv(e), b));
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_symbol(napi_env e, const char* s, napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  v8::Isolate* isolate = v8impl::V8IsolateFromJsEnv(e);
+  if (s == nullptr) {
+    *result = v8impl::JsValueFromV8LocalValue(v8::Symbol::New(isolate));
+  } else {
+    v8::Local<v8::String> string;
+    CHECK_NEW_FROM_UTF8(isolate, string, s);
+
+    *result = v8impl::JsValueFromV8LocalValue(v8::Symbol::New(isolate, string));
+  }
+
+  return GET_RETURN_STATUS();
+}
+
+napi_status napi_create_error(napi_env e, napi_value msg, napi_value* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(result);
+
+  *result = v8impl::JsValueFromV8LocalValue(v8::Exception::Error(
+      v8impl::V8LocalValueFromJsValue(msg).As<v8::String>()));
```

Comment:
Can/should these include actual checks to make sure that `msg` does refer to a string?